File: Examples/test-suite/perl5/bools_runme.pl
use strict;
use warnings;
use Test::More tests => 38;
BEGIN { use_ok('bools') }
require_ok('bools');

my $obj = bools::BoolStructure->new();
isa_ok($obj, 'bools::BoolStructure');

sub check_val { my($ref, $val, $name) = @_;
  is(bools::bo($$ref), $val, "bo($name)");
}
sub check_var { my($ref, $val, $name, $const) = @_;
  is($$ref, $val, $name);
  check_val($ref, $val, $name);
}
sub check_ref { my($ref, $val, $name, $const) = @_;
  unless($const) {
    eval { bools::rbo($$ref) };
    is($@, '', "rbo($name)");
  }
  eval { bools::const_rbo($$ref) };
  is($@, '', "const_rbo($name)");
}
sub check_ptr { my($ref, $val, $name, $const) = @_;
  check_val(\ bools::value($$ref), $val, "value($name)", 1);
  unless($const) {
    eval { bools::pbo($$ref) };
    is($@, '', "pbo($name)");
  }
  eval { bools::const_pbo($$ref) };
  is($@, '', "const_pbo($name)");
  check_ref($ref, $val, $name, $const);
}


check_val(\ 1, 1,  '1', 0);
check_val(\ 0, '', '0', 0);
check_val(\ bools::eax(), 1, 'eax()', 0);
check_var(\ $bools::constbool, '', 'constbool', 1);
check_var(\ $bools::bool1, 1,  'bool1', 0);
check_var(\ $bools::bool2, '', 'bool2', 0);
check_ptr(\ $bools::pbool, 1,  'pbool', 0);
check_ref(\ $bools::rbool, '', 'rbool', 0);
check_ptr(\ $bools::const_pbool, 1,  'pbool', 1);
check_ref(\ $bools::const_rbool, '', 'rbool', 1);
check_var(\ $obj->{m_bool1}, 1,  'obj.m_bool1', 0);
check_var(\ $obj->{m_bool2}, '', 'obj.m_bool2', 0);
check_ptr(\ $obj->{m_pbool}, 1,  'obj.m_pbool', 0);
check_ref(\ $obj->{m_rbool}, '', 'obj.m_rbool', 0);
check_ptr(\ $obj->{m_const_pbool}, 1,  'obj.m_const_pbool', 1);
check_ref(\ $obj->{m_const_rbool}, '', 'obj.m_const_rbool', 1);

File: Examples/test-suite/lua/bools_runme.lua
require("bools")

-- bool constant check
-- TODO: constant should be boolean not number!
assert(bools.constbool ~= true, "constbool")
assert(bools.constbool == 0, "constbool")
-- assert(not bools.constbool, "constbool")

-- bool variables check
assert(bools.bool1, "bool1")
assert(not bools.bool2, "bool2")
assert(bools.value(bools.pbool) == bools.bool1, "bools.pbool")
assert(bools.value(bools.rbool) == bools.bool2, "bools.rbool")
assert(bools.value(bools.const_pbool) == bools.bool1, "const_pbool")
assert(bools.const_rbool == bools.bool2, "const_rbool")

-- bool functions check
assert(not bools.bo(false), "bools.bo(false)")
assert(bools.bo(true), "bools.bo(true)")
assert(bools.value(bools.rbo(bools.rbool)) == bools.value(bools.rbool), "bools.rbo(bools.rbool)")
assert(bools.value(bools.pbo(bools.pbool)) == bools.value(bools.pbool), "bools.pbo(bools.pbool)")
assert(bools.const_rbo(bools.const_rbool) == bools.const_rbool, "const_rbo(bools.const_rbool)")
assert(bools.value(bools.const_pbo(bools.const_pbool)) == bools.value(bools.const_pbool), "bools.const_pbo(bools.const_pbool)")
